<b>Personal Information: </b>

Name: Xiaoyu Yang

e-mail: yangxybupt@outlook.com

Language: Python 3.7

The analysis work is done in Python locally. And Jupyter Notebook makes the analysis work more readable :)

# Overview


Stackoverflow provided me with 2 datasets:  
* posts: Title + specifications of all posts in 2015
* users: List of users with display name, age and location if specified

My task is as an analyst in an online company, I need to help company get some more insights on user behavior and trends on their
platform.

Let's start!

## Analysis for 2 datasets
<b>Note</b>: In this part, I first load the datasets and do some basic statical analysis on them individually, and then join two table to do analysis, such as number of answer counts, tags. And at the same time, I analyze the correlation within features. Finally, I did some prediction and evaluation work of the dataset.

In [1]:
import pandas as pd
import math
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 150)

post_tb = pd.read_csv("stackoverflow2015_posts.csv")
user_tb = pd.read_csv("stackoverflow2015_users.csv")

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2785: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [2]:
post_tb.head()

id                                              title  \
0  34016263               Real alternative for Google Feed API   
1  28321638  Strongly Typed RedirectToAction (Futures) usin...   
2  34054617  Magento 2 installing language packs (nl_NL tra...   
3  27836577  How to mock a tornado coroutine function using...   
4  32064375              Magnet pattern and overloaded methods   

   accepted_answer_id  answer_count  comment_count community_owned_date  \
0                 NaN           0.0             13                  NaN   
1                 NaN           0.0              4                  NaN   
2          34055069.0           1.0              0                  NaN   
3          27880772.0           1.0              0                  NaN   
4          32068645.0           1.0              0                  NaN   

                 creation_date  favorite_count           last_activity_date  \
0  2015-12-01 08:55:53.743 UTC             9.0  2015-12-04 11:52:25.647 UTC   
1  2015-02-04 12:38:37.197 UTC             3.0  2015-02-04 12:38:37.197 UTC   
2  2015-12-02 22:50:38.273 UTC             3.0  2015-12-17 08:06:23.237 UTC   
3  2015-01-08 09:17:39.027 UTC             3.0   2015-01-10 20:15:07.26 UTC   
4  2015-08-18 05:42:56.653 UTC             5.0   2015-08-18 09:31:58.83 UTC   

                last_edit_date last_editor_display_name  last_editor_user_id  \
0  2015-12-04 11:52:25.647 UTC                      NaN            4671020.0   
1                          NaN                      NaN                  NaN   
2  2015-12-11 03:52:32.467 UTC                      NaN            1364007.0   
3    2015-01-08 09:22:43.9 UTC                      NaN             324490.0   
4  2015-08-18 06:16:52.367 UTC                      NaN             554460.0   

  owner_display_name  owner_user_id  parent_id  post_type_id  score  \
0                NaN      4671020.0        NaN             1     15   
1                NaN       826568.0        NaN             1      4   
2                NaN      3215647.0        NaN             1      2   
3                NaN       324490.0        NaN             1      9   
4                NaN       554460.0        NaN             1     13   

                                                tags  view_count  
0            javascript|php|json|rss|google-feed-api      4240.0  
1  c#|asynchronous|asp.net-mvc-5|asp.net-mvc-futures       425.0  
2                       magento|magento2|magento-2.0      2938.0  
3                python|unit-testing|mocking|tornado      2035.0  
4                                     scala|implicit       464.0

In [4]:
user_tb.head()

id   display_name  age                    location
0   431965         Damien  NaN                 Netherlands
1  5503898         Ravers  NaN                    Portugal
2  8823852  Vishnu Baliga  NaN        Kochi, Kerala, India
3  6432429  Matthew Hinea  NaN  Seattle, WA, United States
4  1654841  Milan Thummar  NaN                 Pune, India

In [5]:
post_tb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5594392 entries, 0 to 5594391
Data columns (total 19 columns):
id                          int64
title                       object
accepted_answer_id          float64
answer_count                float64
comment_count               int64
community_owned_date        object
creation_date               object
favorite_count              float64
last_activity_date          object
last_edit_date              object
last_editor_display_name    object
last_editor_user_id         float64
owner_display_name          object
owner_user_id               float64
parent_id                   float64
post_type_id                int64
score                       int64
tags                        object
view_count                  float64
dtypes: float64(7), int64(4), object(8)
memory usage: 811.0+ MB


In [6]:
user_tb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10097978 entries, 0 to 10097977
Data columns (total 4 columns):
id              int64
display_name    object
age             float64
location        object
dtypes: float64(1), int64(1), object(2)
memory usage: 308.2+ MB


In [7]:
post_tb.columns.values.tolist()

['id',
 'title',
 'accepted_answer_id',
 'answer_count',
 'comment_count',
 'community_owned_date',
 'creation_date',
 'favorite_count',
 'last_activity_date',
 'last_edit_date',
 'last_editor_display_name',
 'last_editor_user_id',
 'owner_display_name',
 'owner_user_id',
 'parent_id',
 'post_type_id',
 'score',
 'tags',
 'view_count']

In [8]:
user_tb.columns.values.tolist()

['id', 'display_name', 'age', 'location']

<b>Note: </b> 

From the Part above, we know that Posts has 5594392 records and User has 10097978 results. And the column names are listed in above two cells. So We start analyzing user table individually.

In [10]:
len(user_tb[user_tb["age"].notna()])
# number of nan value in age

0

In [11]:
len(user_tb[user_tb["location"].notna()])
# number of nan value in location

2497373

In [12]:
len(user_tb[user_tb["id"].notna()])
# number of nan value in id

10097978

In [14]:
len(set(user_tb["id"].values.tolist()))
# check distinct number of id

10097978

In [15]:
2497373/10097978

0.24731416527150288

<b>Note:</b>

<b>From the above code we could know that</b> 

* all age information is missing
* about 24.7% information is missing
* each user has an individual id, which means his/her id is distinct to the others
* we do not consider display name information in this experiment

We drop age and display name now for further process

In [3]:
user_tb.drop(["age","display_name"] , axis = 1, inplace = True)

In [4]:
user_tb.head()

id                    location
0   431965                 Netherlands
1  5503898                    Portugal
2  8823852        Kochi, Kerala, India
3  6432429  Seattle, WA, United States
4  1654841                 Pune, India

<b>Note:</b>

Then, we want to know the distribution of user's location, so we first count the values of location

In [20]:
user_tb.location.value_counts()

India                                                                             57800
Bangalore, Karnataka, India                                                       41231
Germany                                                                           28234
Hyderabad, Telangana, India                                                       24238
Pune, Maharashtra, India                                                          23084
Chennai, Tamil Nadu, India                                                        21337
China                                                                             20185
London, United Kingdom                                                            19830
United States                                                                     17404
France                                                                            16674
Mumbai, Maharashtra, India                                                        14912
Paris, France                   

<b>Note:</b>

We know the differnet location is 173323, but some people regisiter their information to cities, some are only countries, and the spelling are also the problem, for example, some people use "China" but some use "china". We want to learn about the distribution of countries. Country is always the last word of location information.

So we only keep the countries information and change all digit to capital and do the analysis work

In [5]:
def check_last_word(total_location):
    multiple_str = str(total_location).split(",")
    return multiple_str[-1].strip().upper()
user_tb['location'] = user_tb['location'].apply(lambda x:check_last_word(x))

In [24]:
user_tb.location.value_counts()

NAN                                                                                                     7600609
INDIA                                                                                                    396307
UNITED STATES                                                                                            195916
USA                                                                                                      134703
UNITED KINGDOM                                                                                            66256
CANADA                                                                                                    61723
GERMANY                                                                                                   57228
FRANCE                                                                                                    50314
CHINA                                                                                                   

In [26]:
(396307+195916+134703+66256+61723+57228+50314+47572+39195+37549+33504+30101)/(10097978-7600609)

0.4606319690842643

<b>Note:</b>

We could know that Except for missing values the top 10 user comes from, about 46% of the recorded users:

India, United States/USA, United Kingdom/UK, Canada, Germany, France, China, Australia, PAKISTAN, BRASIL


So the company should put more information on business of the district above

We the focus on post information

Before we move on to next step. Because we have already analysis the user table, we left join two tables on "owner_user_id". (There are 2 user id in posts columns, the othe is "last_editor_user_id", but last editor could be randomly, and it is more meaningful to analysis on owner user behavior)


In [6]:
def to_int(x):
    if math.isnan(x):
        return -1
    else:
        return int(x)
# owner user is float, so we need to transfer it into int

In [7]:
post_tb['owner_user_id2'] = post_tb['owner_user_id'].apply(lambda x:to_int(x))
user_tb.rename(columns = {'id':'owner_user_id2'},inplace = True)
post_tb = pd.merge(post_tb,user_tb,how = 'left',on = 'owner_user_id2')

In [8]:
post_tb.drop(["owner_user_id2"] , axis = 1, inplace = True)
post_tb
# .drop(["Unnamed: 0"] , axis = 1, inplace = True)

id                                              title  \
0        34016263               Real alternative for Google Feed API   
1        28321638  Strongly Typed RedirectToAction (Futures) usin...   
2        34054617  Magento 2 installing language packs (nl_NL tra...   
3        27836577  How to mock a tornado coroutine function using...   
4        32064375              Magnet pattern and overloaded methods   
5        32425466    Load More After Coming to Bottom of UITableView   
6        30952612  Include and reference an image in react with w...   
7        34498830     Decode X-JWT-Assertion using axiom-api in java   
8        30383404  What is data type for Python Keras deep learni...   
9        30714031                   Tuple hell with many Slick joins   
10       30474538  Possible to show console messages (written wit...   
11       34546446       Flattening TypeScript typings or interfaces?   
12       28758173  Integrate Swift framework to objective-c appli...   
13       27826508  Create all possible combiations of 01 or 2 1 s...   
14       28674344  Windows Service fails to start interactive pro...   
15       30523118  How to Upload gif files to parse in android as...   
16       33034930  How to use angular2 DynamicComponentLoader in ...   
17       28917651  Building a Cocoapod with Swift and dependency ...   
18       28949722  Android tests fail on Travis with ShellCommand...   
19       28108107  How to create a Row from a List or Array in Sp...   
20       32973743  Making Bootstrap modal resizable with jQuery U...   
21       27762236  Line Breaks and Number of Lines in Swift Label...   
22       31402531  Google IAP returns short purchase token for ve...   
23       31377629                  Git always ask passphrase for key   
24       30109169               Create FullPage Screenshot WebDriver   
25       31353299  Swift Dictionary slow even with optimizations:...   
26       30473652             Cordova/Phonegap iOS Parse-Push Plugin   
27       32419510           How to get reproducible results in keras   
28       31346267  What is the need and use of <mvc:default-servl...   
29       32822101  How to programmatically open the Permission Sc...   
30       29090692  How is ambiguity determined in the overload re...   
31       32955560  Difference between '01' and '1' in a JavaScrip...   
32       29361552  Android List View with BaseAdapter Drag and Dr...   
33       28270282  What is the best way to check if a UIAlertCont...   
34       28765806  Existing Google Analytics events and Google Ta...   
35       28359851         how to install android studio full offline   
36       29653927  Green images when doing a JPEG encoding from Y...   
37       28932281  What is wrong with my SQL here? #1089 - Incorr...   
38       28338935                                                NaN   
39       30407942                                                NaN   
40       28959043                                                NaN   
41       31117150                                                NaN   
42       27875362                                                NaN   
43       29449221                                                NaN   
44       29609368                                                NaN   
45       29967813                                                NaN   
46       33385549                                                NaN   
47       27849234                                                NaN   
48       32202224                                                NaN   
49       28723575                                                NaN   
50       29045865                                                NaN   
51       30815435                                                NaN   
52       28740787                                                NaN   
53       30400863                                                NaN   
54       30259051                                      

<b>Note:</b>

We first do statistical work on "answer_count", "comment_count", "favorite_count" and "view_count"

For "answer_count", it is also good to analysis it with accepted situation, i.e. accepted_answer_id

In [40]:
def cal_per(df):
    return len(df)/5594392
print("No anwser ratio "+str(cal_per(post_tb[post_tb["answer_count"].isnull()])))
print("1 anwser ratio "+str(cal_per(post_tb[post_tb["answer_count"]==1])))
print("2 anwser ratio "+str(cal_per(post_tb[post_tb["answer_count"]==2])))
print("0 anwser ratio "+str(cal_per(post_tb[post_tb["answer_count"]==0])))
print("More than 2 anwser ratio "+str(cal_per(post_tb[post_tb["answer_count"]>2])))

No anwser ratio 0.5730241999488058
1 anwser ratio 0.21894926204670678
2 anwser ratio 0.08189504775496605
0 anwser ratio 0.08764455547626981
More than 2 anwser ratio 0.0384869347732515


We could see that about 57.3% posts do not have anwser and 8.7% posts have 0 anwser

The most anwser have 1 or 2 anwser, about 30% of the results

In [46]:
print("Have more than 0 anwser but not accepted "+str(cal_per(post_tb[post_tb["accepted_answer_id"].isnull() & post_tb["answer_count"].notna()& post_tb["answer_count"]!=0])))

Have more than 0 anwser but not accepted 0.14707603614476783


We could see about 14.7% of the posts have anwsers, and anwser is not 0, but no anwsers. Company should focus on this question and find out the solution of anwser acceptance

In [48]:
post_tb[["id","answer_count"]].groupby("answer_count").count().reset_index()

answer_count       id
0            0.0   490318
1            1.0  1224888
2            2.0   458153
3            3.0   144901
4            4.0    46648
5            5.0    15374
6            6.0     5102
7            7.0     1853
8            8.0      713
9            9.0      300
10          10.0      161
11          11.0       95
12          12.0       52
13          13.0       38
14          14.0       24
15          15.0       17
16          16.0       11
17          17.0        3
18          18.0        2
19          19.0        2
20          20.0        5
21          21.0        1
22          22.0        2
23          23.0        1
24          24.0        1
25          25.0        2
26          26.0        1
27          27.0        2

We could see that highest answer number is 27 and lowest is 0

In [51]:
import numpy as np

In [54]:
np.median(post_tb[post_tb["answer_count"].notna()]["answer_count"].values.tolist())

1.0

The median value is 1, which proved the preceding result of distribution

Same method, for the "comment_count"

In [55]:
post_tb[["id","comment_count"]].groupby("comment_count").count().reset_index()

comment_count       id
0               0  2624099
1               1   923535
2               2   703548
3               3   428850
4               4   296864
5               5   191679
6               6   133320
7               7    90256
8               8    61775
9               9    41535
10             10    28375
11             11    19483
12             12    13871
13             13     9703
14             14     6978
15             15     5009
16             16     3832
17             17     2751
18             18     2170
19             19     1604
20             20     1172
21             21      857
22             22      650
23             23      513
24             24      423
25             25      290
26             26      247
27             27      182
28             28      147
29             29      115
30             30       89
31             31       79
32             32       70
33             33       46
34             34       38
35             35       34
36             36       23
37             37       24
38             38       18
39             39       11
40             40       14
41             41       11
42             42       16
43             43       11
44             44        9
45             45        4
46             46        7
47             47        3
48             48        9
49             49        4
50             50        2
51             51        2
52             52        3
53             53        4
54             54        2
55             55        4
56             56        1
57             57        5
58             58        2
59             59        1
60             60        1
61             64        2
62             69        1
63             70        1
64             72        1
65             75        1
66             78        3
67             91        1
68            119        2

In [56]:
np.median(post_tb[post_tb["comment_count"].notna()]["comment_count"].values.tolist())

1.0

In [58]:
post_tb[post_tb["comment_count"]==119]

id                                          title  \
1335949  29479702                                            NaN   
1813173  32240629  UIImageView Frame Doesn't Reflect Constraints   

         accepted_answer_id  answer_count  comment_count  \
1335949                 NaN           NaN            119   
1813173          32261177.0           1.0            119   

                community_owned_date                creation_date  \
1335949  2015-04-06 21:07:14.203 UTC  2015-04-06 21:07:14.203 UTC   
1813173                          NaN  2015-08-27 04:17:12.267 UTC   

         favorite_count           last_activity_date  \
1335949             NaN  2015-04-17 19:32:04.247 UTC   
1813173             NaN  2015-08-27 23:45:54.283 UTC   

                      last_edit_date last_editor_display_name  \
1335949  2015-04-17 19:32:04.247 UTC                      NaN   
1813173  2015-08-27 04:50:07.493 UTC                      NaN   

         last_editor_user_id owner_display_name  owner_user_id  parent_id  \
1335949                811.0                NaN       100297.0        NaN   
1813173            2809263.0                NaN      2809263.0        NaN   

         post_type_id  score                                     tags  \
1335949             6  10893                                      NaN   
1813173             1      0  ios|swift|uiview|autolayout|constraints   

         view_count location  
1335949         NaN       UK  
1813173       113.0      NAN

The median value of "comment_count" is also 1, and it is mainly distributed in 0 to 10 comments for each anwser. The largest value is 119, mucher higher than the others. So company should make survey on these kind of posts to attracts visitors' comment.

Then for "favorite_count"

In [59]:
post_tb[["id","favorite_count"]].groupby("favorite_count").count().reset_index()

favorite_count      id
0              0.0   70861
1              1.0  216593
2              2.0   36500
3              3.0    9625
4              4.0    3588
5              5.0    1744
6              6.0    1002
7              7.0     605
8              8.0     400
9              9.0     268
10            10.0     200
11            11.0     162
12            12.0     123
13            13.0     105
14            14.0      69
15            15.0      50
16            16.0      50
17            17.0      43
18            18.0      39
19            19.0      39
20            20.0      21
21            21.0      24
22            22.0      25
23            23.0      24
24            24.0      11
25            25.0      11
26            26.0      12
27            27.0      13
28            28.0       7
29            29.0       8
30            30.0       6
31            31.0      12
32            32.0       6
33            33.0      13
34            34.0       1
35            35.0       8
36            36.0       5
37            37.0       6
38            38.0       5
39            39.0       8
40            40.0       4
41            41.0       3
42            42.0       2
43            43.0       2
44            44.0       1
45            45.0       5
46            46.0       3
47            47.0       4
48            48.0       3
49            49.0       5
50            51.0       3
51            52.0       3
52            54.0       3
53            55.0       4
54            57.0       1
55            58.0       1
56            60.0       2
57            61.0       2
58            62.0       2
59            63.0       1
60            66.0       1
61            67.0       2
62            71.0       1
63            73.0       3
64            74.0       2
65            76.0       1
66            77.0       2
67            78.0       1
68            80.0       1
69            81.0       1
70            82.0       1
71            83.0       2
72            85.0       2
73            89.0       1
74            90.0       1
75            95.0       1
76           109.0       1
77           125.0       1
78           134.0       1
79           147.0       1
80           158.0       1
81           172.0       1
82           189.0       1
83           205.0       1
84           236.0       1
85           328.0       1

In [61]:
np.median(post_tb[post_tb["favorite_count"].notna()]["favorite_count"].values.tolist())

1.0

The median value of "favorite_count" is also 1, and it is mainly distributed in 0 to 10 favorites for each anwser. The largest value is 205, 236, 328, mucher higher than the others. So company should make survey on these kind of posts to attracts visitors' comment.

And a different thing is that favourite count of 1 is much higher than 0, maybe that means for the creator they always favourite their own questions :), it is better for company do survey on it.

Then for "view_count"

In [62]:
post_tb[["id","view_count"]].groupby("view_count").count().reset_index()

view_count     id
0             2.0     26
1             3.0    100
2             4.0    274
3             5.0    534
4             6.0    822
5             7.0   1353
6             8.0   1888
7             9.0   2531
8            10.0   3321
9            11.0   4125
10           12.0   5124
11           13.0   5886
12           14.0   7093
13           15.0   8172
14           16.0   9398
15           17.0  10466
16           18.0  11708
17           19.0  12749
18           20.0  13795
19           21.0  15279
20           22.0  16135
21           23.0  17487
22           24.0  18162
23           25.0  19130
24           26.0  20008
25           27.0  20627
26           28.0  21344
27           29.0  21911
28           30.0  22645
29           31.0  22866
30           32.0  23023
31           33.0  23643
32           34.0  23879
33           35.0  23741
34           36.0  24045
35           37.0  23949
36           38.0  23992
37           39.0  23725
38           40.0  23709
39           41.0  23352
40           42.0  23217
41           43.0  22812
42           44.0  22475
43           45.0  22877
44           46.0  22096
45           47.0  21635
46           48.0  21684
47           49.0  21072
48           50.0  20829
49           51.0  20260
50           52.0  20080
51           53.0  19767
52           54.0  19392
53           55.0  19150
54           56.0  18713
55           57.0  18306
56           58.0  18117
57           59.0  17662
58           60.0  17152
59           61.0  17150
60           62.0  16488
61           63.0  16171
62           64.0  15927
63           65.0  15727
64           66.0  15440
65           67.0  14926
66           68.0  14548
67           69.0  14309
68           70.0  14121
69           71.0  13887
70           72.0  13473
71           73.0  13169
72           74.0  12979
73           75.0  12488
74           76.0  12448
...           ...    ...
10676     62471.0      1
10677     62835.0      1
10678     63228.0      1
10679     63250.0      1
10680     63467.0      1
10681     63607.0      1
10682     63663.0      1
10683     63669.0      1
10684     64603.0      1
10685     64936.0      1
10686     65084.0      1
10687     65887.0      1
10688     66626.0      1
10689     67048.0      1
10690     67177.0      1
10691     67417.0      1
10692     67974.0      1
10693     68151.0      1
10694     68496.0      1
10695     69286.0      1
10696     69720.0      1
10697     70166.0      1
10698     70744.0      1
10699     71410.0      1
10700     72188.0      1
10701     72592.0      1
10702     72676.0      1
10703     72971.0      1
10704     73503.0      1
10705     74976.0      1
10706     75201.0      1
10707     75255.0      1
10708     75910.0      1
10709     76418.0      1
10710     76627.0      1
10711     77213.0      1
10712     78031.0      1
10713     78323.0      2
10714     78344.0      1
10715     79381.0      1
10716     80931.0      1
10717     82232.0      1
10718     83652.0      1
10719     85359.0      1
10720     85746.0      1
10721     86656.0      1
10722     88679.0      1
10723     89899.0      1
10724     89955.0      1
10725     90127.0      1
10726     93937.0      1
10727     95485.0      1
10728     97237.0      1
10729     98534.0      1
10730     99141.0      1
10731    101461.0      1
10732    104562.0      1
10733    105932.0      1
10734    107277.0      1
10735    111740.0      1
10736    112842.0      1
10737    114519.0      1
10738    125813.0      1
10739    126210.0      1
10740    128451.0      1
10741    135616.0      1
10742    137771.0      1
10743    158464.0      1
10744    177337.0      1
10745    218715.0      1
10746    235734.0      1
10747    251252.0      1
10748    266043.0      1
10749    267778.0      1
10750    318907.0      1

[10751 rows x 2 columns]

In [63]:
np.median(post_tb[post_tb["view_count"].notna()]["view_count"].values.tolist())

78.0

In [67]:
from scipy.stats import normaltest
normaltest(post_tb[post_tb["view_count"].notna()]["view_count"].values.tolist())

NormaltestResult(statistic=8566354.999981645, pvalue=0.0)

The median value of "view_count" is also 78, it is main distributed around this value. After performing normal test, it does not obey such test.

Then we analysis the correlation within these values

In [69]:
post_tb[["answer_count","comment_count","favorite_count","view_count"]].corr()

answer_count  comment_count  favorite_count  view_count
answer_count        1.000000       0.040317        0.205724    0.173082
comment_count       0.040317       1.000000        0.047302    0.010082
favorite_count      0.205724       0.047302        1.000000    0.507507
view_count          0.173082       0.010082        0.507507    1.000000

From the correlation matrix we could see that answer_count and favorite_count, answer_count and view_count, favorite_count and view_count has higher correlation, which means when one feature, such as favorite_count increase for some specific post, the other such as answer_count has higer possibility of increasing

The analysis of this part is not done yet, I will do further analysis combined with "tages" in parts later

In this part we want to see the score information of the posts

In [72]:
post_tb[["id","score"]].groupby("score").count().reset_index()

score       id
0      -63        1
1      -50        1
2      -42        1
3      -40        1
4      -30        1
5      -28        2
6      -25        1
7      -21        6
8      -20        7
9      -19        2
10     -18        5
11     -17       11
12     -16        7
13     -15       11
14     -14       15
15     -13       38
16     -12       43
17     -11      109
18     -10      162
19      -9      337
20      -8      593
21      -7     1215
22      -6     2442
23      -5     5479
24      -4    11625
25      -3    23814
26      -2    55464
27      -1   166098
28       0  2775645
29       1  1419616
30       2   575978
31       3   248780
32       4   118347
33       5    60581
34       6    35577
35       7    22079
36       8    14732
37       9    10164
38      10     7604
39      11     5679
40      12     4511
41      13     3611
42      14     2851
43      15     2362
44      16     1954
45      17     1630
46      18     1418
47      19     1183
48      20     1099
49      21      934
50      22      794
51      23      665
52      24      653
53      25      546
54      26      527
55      27      435
56      28      427
57      29      385
58      30      343
59      31      304
60      32      296
61      33      275
62      34      260
63      35      216
64      36      193
65      37      203
66      38      183
67      39      137
68      40      148
69      41      123
70      42      140
71      43      139
72      44      106
73      45      118
74      46      105
..     ...      ...
320    412        1
321    414        1
322    415        1
323    428        2
324    431        1
325    439        1
326    440        1
327    446        1
328    449        1
329    451        1
330    455        1
331    478        1
332    490        1
333    501        1
334    508        1
335    513        1
336    517        1
337    520        1
338    524        1
339    526        2
340    548        1
341    554        1
342    572        1
343    588        1
344    591        1
345    593        1
346    605        1
347    608        1
348    653        1
349    654        1
350    659        1
351    671        1
352    679        1
353    680        1
354    681        1
355    685        1
356    691        1
357    698        1
358    724        1
359    769        1
360    815        1
361    884        1
362    945        1
363   1018        1
364   1078        1
365   1223        1
366   1406        1
367   1492        1
368   1617        1
369   1745        1
370   1820        1
371   1998        1
372   2483        1
373   2536        1
374   2610        1
375   2646        2
376   2662        1
377   3058        1
378   3123        1
379   3173        1
380   3297        1
381   3899        1
382   4126        1
383   4353        1
384   4593        1
385   5190        1
386   5204        1
387   5455        1
388   5656        1
389   5915        1
390   7302        1
391   7387        1
392   9368        1
393  10893        1
394  11941        1

[395 rows x 2 columns]

We could see that the score range is from -63 to 11941, the interesting is that we got negative value of score, so I'd like to first calculate the negative value ratio.

In [73]:
neg_num = 0
for each in post_tb.score.values.tolist():
    if each<0:
        neg_num+=1
neg_num/len(post_tb.score.values.tolist())

0.04781413243834182

In [77]:
post_tb[post_tb["score"]<0].head()

id title  accepted_answer_id  answer_count  comment_count  \
677  29559861   NaN                 NaN           NaN              0   
678  31080771   NaN                 NaN           NaN              0   
679  28583045   NaN                 NaN           NaN              0   
680  29271828   NaN                 NaN           NaN              0   
681  32706565   NaN                 NaN           NaN              0   

    community_owned_date                creation_date  favorite_count  \
677                  NaN  2015-04-10 11:06:32.183 UTC             NaN   
678                  NaN  2015-06-26 19:28:13.977 UTC             NaN   
679                  NaN  2015-02-18 11:55:28.147 UTC             NaN   
680                  NaN   2015-03-26 05:44:50.14 UTC             NaN   
681                  NaN  2015-09-22 00:32:53.137 UTC             NaN   

              last_activity_date               last_edit_date  \
677  2015-04-11 11:06:45.757 UTC  2015-04-11 11:06:45.757 UTC   
678  2015-06-26 20:50:26.967 UTC  2015-06-26 20:50:26.967 UTC   
679  2015-12-01 12:41:13.413 UTC  2015-12-01 12:41:13.413 UTC   
680  2015-03-28 05:40:10.603 UTC  2015-03-28 05:40:10.603 UTC   
681  2015-09-22 01:05:18.997 UTC  2015-09-22 01:05:18.997 UTC   

    last_editor_display_name  last_editor_user_id owner_display_name  \
677                      NaN            1401962.0                NaN   
678                      NaN            1386703.0                NaN   
679                      NaN             316424.0                NaN   
680                      NaN            4550791.0                NaN   
681                      NaN            4526552.0                NaN   

     owner_user_id   parent_id  post_type_id  score tags  view_count location  
677      4430741.0    258626.0             2     -1  NaN         NaN      NAN  
678      5054279.0  31077317.0             2     -1  NaN         NaN      NAN  
679      4540247.0  11169281.0             2     -1  NaN         NaN      NAN  
680      4550791.0  29271026.0             2     -1  NaN         NaN      NAN  
681      4526552.0  32706381.0             2     -1  NaN         NaN      NAN

We could see that about 4.78% of the posts holds the negative score. So company should focus on the negative score posts to improve posts services

Then we want to know the inforamtion of post type of each post

In [78]:
post_tb[["id","post_type_id"]].groupby("post_type_id").count().reset_index()

post_type_id       id
0             1  2388670
1             2  3194665
2             4     5496
3             5     5496
4             6       65

We could know that the main part of post type is 1 and 2, we donot have type3, and the rest is 4, 5, 6. Company could focus on improving services of post type 2 and 1 first. Also, it may be a good method to eccorage people to use 4, 5, 6 to post.

Then because a very interesting information we have is that we have date information of each post in year 2015, so I want to make clear the post distribution of Month, Hour and Weekday, and important, the times interval between created date and last activity date.

In [9]:
post_tb['creation_date'] = post_tb['creation_date'].apply(lambda x:x.replace(' UTC','+0000'))
post_tb['last_activity_date'] = post_tb['last_activity_date'].apply(lambda x:x.replace(' UTC','+0000'))
post_tb['creation_month'] = post_tb['creation_date'].apply(lambda x:x[5:7])
post_tb['creation_hour'] = post_tb['creation_date'].apply(lambda x:x[10:13])
post_tb['time_interval'] = pd.DataFrame(pd.to_datetime(post_tb['last_activity_date']) - pd.to_datetime(post_tb['creation_date']))
post_tb['weekday'] = post_tb['creation_date'].apply(lambda x:pd.to_datetime(x).weekday())

In [10]:
post_tb.drop(["creation_date","last_activity_date"] , axis = 1, inplace = True)

We drop duplicated information here

In [12]:
post_tb['time_interval'] = post_tb['time_interval'].apply(lambda x: x.days)

In [13]:
post_tb.head()

id                                              title  \
0  34016263               Real alternative for Google Feed API   
1  28321638  Strongly Typed RedirectToAction (Futures) usin...   
2  34054617  Magento 2 installing language packs (nl_NL tra...   
3  27836577  How to mock a tornado coroutine function using...   
4  32064375              Magnet pattern and overloaded methods   

   accepted_answer_id  answer_count  comment_count community_owned_date  \
0                 NaN           0.0             13                  NaN   
1                 NaN           0.0              4                  NaN   
2          34055069.0           1.0              0                  NaN   
3          27880772.0           1.0              0                  NaN   
4          32068645.0           1.0              0                  NaN   

   favorite_count               last_edit_date last_editor_display_name  \
0             9.0  2015-12-04 11:52:25.647 UTC                      NaN   
1             3.0                          NaN                      NaN   
2             3.0  2015-12-11 03:52:32.467 UTC                      NaN   
3             3.0    2015-01-08 09:22:43.9 UTC                      NaN   
4             5.0  2015-08-18 06:16:52.367 UTC                      NaN   

   last_editor_user_id owner_display_name  owner_user_id  parent_id  \
0            4671020.0                NaN      4671020.0        NaN   
1                  NaN                NaN       826568.0        NaN   
2            1364007.0                NaN      3215647.0        NaN   
3             324490.0                NaN       324490.0        NaN   
4             554460.0                NaN       554460.0        NaN   

   post_type_id  score                                               tags  \
0             1     15            javascript|php|json|rss|google-feed-api   
1             1      4  c#|asynchronous|asp.net-mvc-5|asp.net-mvc-futures   
2             1      2                       magento|magento2|magento-2.0   
3             1      9                python|unit-testing|mocking|tornado   
4             1     13                                     scala|implicit   

   view_count     location creation_month creation_hour  time_interval  \
0      4240.0          NAN             12            08              3   
1       425.0      URUGUAY             02            12              0   
2      2938.0  NETHERLANDS             12            22             14   
3      2035.0       TAIWAN             01            09              2   
4       464.0           CA             08            05              0   

   weekday  
0        1  
1        2  
2        2  
3        3  
4        1

We first have a look at the distribution between months

In [14]:
post_tb[["id","creation_month"]].groupby("creation_month").count().reset_index()

creation_month      id
0              01  417460
1              02  420321
2              03  469464
3              04  482161
4              05  468742
5              06  483342
6              07  511968
7              08  475543
8              09  466241
9              10  485682
10             11  461953
11             12  451515

We could know that the posts in summer (month: 6 7 8 9) is average 20% higer than winter (month 1 2) per month

Then analysis for the hour

In [15]:
post_tb[["id","creation_hour"]].groupby("creation_hour").count().reset_index()

creation_hour      id
0             00  132979
1             01  123885
2             02  124478
3             03  126660
4             04  139533
5             05  169136
6             06  209864
7             07  239051
8             08  255611
9             09  296306
10            10  299205
11            11  293364
12            12  300139
13            13  317075
14            14  326270
15            15  320099
16            16  295824
17            17  275019
18            18  268006
19            19  258358
20            20  248701
21            21  226252
22            22  190276
23            23  158301

We could know that from hour 0 to hour 23, the peak hour of creatign posts is from 9 to 21, which is much higer than in the evening and mid-night. Users tend to post posts in day. So company should pay more attention to day time services.

Then analysis for the weekday

In [16]:
post_tb[["id","weekday"]].groupby("weekday").count().reset_index()

weekday      id
0        0  871751
1        1  962993
2        2  983404
3        3  967650
4        4  858746
5        5  476298
6        6  473550

N.B. in the weekday calculated by python, 0 means Monday, 1 means Tuesday and etc.

So we could know that in weekend (5 to Saturday and 6 to Sunday), the number of posts is much lower than that in weekday, which is only half of the weekday. Weekday posts stay stable. Maybe because people always meet problems and posts them during work time.

Then for time interval

In [17]:
post_tb[["id","time_interval"]].groupby("time_interval").count().reset_index()

time_interval       id
0                0  4832518
1                1   149846
2                2    80464
3                3    43209
4                4    30188
5                5    23481
6                6    19313
7                7    15121
8                8    11863
9                9     9957
10              10     8342
11              11     7675
12              12     7206
13              13     6950
14              14     6123
15              15     5009
16              16     4441
17              17     4180
18              18     4084
19              19     3932
20              20     3982
21              21     3767
22              22     3394
23              23     2923
24              24     2839
25              25     2863
26              26     2759
27              27     2968
28              28     2978
29              29     2557
30              30     2308
31              31     2099
32              32     2124
33              33     2256
34              34     2327
35              35     2254
36              36     1981
37              37     1850
38              38     1720
39              39     1721
40              40     1764
41              41     1954
42              42     1785
43              43     1719
44              44     1548
45              45     1553
46              46     1550
47              47     1519
48              48     1661
49              49     1696
50              50     1488
51              51     1380
52              52     1405
53              53     1383
54              54     1422
55              55     1570
56              56     1597
57              57     1458
58              58     1319
59              59     1326
60              60     1308
61              61     1320
62              62     1447
63              63     1442
64              64     1286
65              65     1234
66              66     1187
67              67     1177
68              68     1259
69              69     1336
70              70     1331
71              71     1274
72              72     1156
73              73     1141
74              74     1079
..             ...      ...
452            452      114
453            453      126
454            454      128
455            455      128
456            456      127
457            457       98
458            458       82
459            459      123
460            460      119
461            461      105
462            462      120
463            463       88
464            464      100
465            465       84
466            466       85
467            467      107
468            468      104
469            469       97
470            470       75
471            471       70
472            472       68
473            473       71
474            474       82
475            475       88
476            476       76
477            477       70
478            478       63
479            479       57
480            480       74
481            481       82
482            482       70
483            483       72
484            484       46
485            485       66
486            486       42
487            487       57
488            488       57
489            489       68
490            490       65
491            491       73
492            492       50
493            493       61
494            494       52
495            495       52
496            496       47
497            497       39
498            498       36
499            499       39
500            500       37
501            501       44
502            502       40
503            503       33
504            504       31
505            505       29
506            506       26
507            507       32
508            508       18
509            509       31
510            510       28
511            511       29
512            512       22
513            513       17
514            514       12
515            515       16
516            516       14
517            517       16
5

In [18]:
4832518/5594392
# ratio of 0 days

0.8638146915697005

In [22]:
sum(post_tb[["id","time_interval"]].groupby("time_interval").count().reset_index()["id"].values.tolist()[:10])/5594392
# ratio of 0 to 9 days

0.9323551156229309

Although there are many posts activity last a long time, we could know that most posts last activity is within only 1 day, the ratio is about 86.4%, and the posts last activity within 10 days ratio is about 93.2%. Company should focus more on the current day posts services.

Then a important information is about the owner user, we want to find out for each user, how many posts has he/she posted and where they come from.

In [25]:
owner_small = post_tb[["owner_user_id","id"]].groupby("owner_user_id").count().reset_index().sort_values(by = "id",ascending = False)
owner_small

owner_user_id    id
133021       1144035.0  7570
60290         548225.0  4300
463552       3732271.0  3392
407864       3297613.0  3110
182241       1491895.0  2954
11559         100297.0  2866
477590       3832970.0  2800
266531       2141635.0  2536
84434         771848.0  2467
12984         114251.0  2413
13067         115145.0  2372
1430            6309.0  2357
17498         157247.0  2326
144265       1221571.0  2306
47683         434551.0  2201
36779         335858.0  2090
23004         207421.0  2005
3773           22656.0  1965
111231        992484.0  1944
3319           19068.0  1910
37430         341994.0  1889
87236         795990.0  1853
300593       2422776.0  1841
0                 -1.0  1791
89726         816620.0  1719
50996         462627.0  1702
302188       2435473.0  1698
507544       4039065.0  1669
183548       1501794.0  1580
251597       2025923.0  1556
356634       2877241.0  1555
55831         505088.0  1553
105273        949476.0  1540
456614       3679490.0  1530
11925         104349.0  1494
14898         131872.0  1492
355520       2867928.0  1490
62867         571407.0  1480
203187       1651298.0  1469
218316       1766831.0  1455
347462       2801559.0  1453
270423       2173917.0  1444
257236       2069064.0  1435
31380         284240.0  1377
197190       1603351.0  1359
368024       2970947.0  1346
119394       1048572.0  1305
8064           65863.0  1274
77398         704848.0  1245
10046          85371.0  1236
184031       1505120.0  1231
57366         519413.0  1220
48365         440558.0  1199
231040       1863229.0  1178
440852       3559349.0  1171
137677       1175966.0  1166
246228       1983495.0  1164
57654         522444.0  1139
1624            7432.0  1126
492134       3933332.0  1119
2730           14860.0  1108
450950       3636601.0  1104
414247       3349551.0  1098
187367       1529630.0  1089
22759         205185.0  1087
61694         560648.0  1078
500258       3989608.0  1076
373240       3016153.0  1073
8391           69083.0  1071
2766           15168.0  1048
55469         501696.0  1040
439030       3545273.0  1039
272312       2189127.0  1029
15724         139985.0  1025
347530       2802040.0  1022
...                ...   ...
63158         573991.0     1
569967       4410306.0     1
569963       4410288.0     1
569961       4410280.0     1
569901       4410096.0     1
569900       4410095.0     1
569899       4410092.0     1
63198         574292.0     1
569819       4409855.0     1
569820       4409856.0     1
569822       4409858.0     1
569825       4409869.0     1
569826       4409871.0     1
569827       4409881.0     1
569829       4409886.0     1
569843       4409949.0     1
569830       4409887.0     1
63197         574279.0     1
569832       4409909.0     1
569838       4409934.0     1
569839       4409938.0     1
569840       4409940.0     1
569818       4409854.0     1
569817       4409853.0     1
569816       4409852.0     1
569812       4409840.0     1
569811       4409838.0     1
63204         574379.0     1
569806       4409821.0     1
569805       4409819.0     1
63206         574426.0     1
569802       4409803.0     1
569800       4409800.0     1
569799       4409799.0     1
569797       4409790.0     1
569795       4409782.0     1
569792       4409770.0     1
569791       4409767.0     1
569789       4409761.0     1
569841       4409945.0     1
569844       4409952.0     1
569898       4410091.0     1
569884       4410058.0     1
569873       4410016.0     1
569877       4410040.0     1
569878       4410044.0     1
569880       4410052.0     1
569881       4410055.0     1
569883       4410057.0     1
569885       4410060.0     1
569845       4409954.0     1
569887       4410067.0     1
63175         574099.0     1
569892       4410080.0     1
63174         574093.0     1
569896       4410087.0     1
569897       4410088.0     1
569868       4410010.0     1
63180         574130.0     1
63183         574154.0     1
569862       4409992.0     1
63185  

N.B. in the data processing procedure, if a post have no user id, I change it to -1.

From the table above, we could see that a lot user have thousands of value of posts, the most activite user is user 1144035 who has 7570 posts. Is that means all user are active?

So then I calculate the number of users who have post less than 10

In [30]:
cal_lst2 = [0,0,0,0,0,0,0,0,0]
cal_large = 0
for each in list(owner_small['id']):
    if each ==1:
        cal_lst2[0]+=1
    if each ==2:
        cal_lst2[1]+=1
    if each ==3:
        cal_lst2[2]+=1
    if each ==4:
        cal_lst2[3]+=1
    if each ==5:
        cal_lst2[4]+=1
    if each ==6:
        cal_lst2[5]+=1
    if each ==7:
        cal_lst2[6]+=1
    if each ==8:
        cal_lst2[7]+=1
    if each ==9:
        cal_lst2[8]+=1
    if each > 100:
        cal_large +=1
cal_lst2

[478164, 174631, 89287, 56113, 38340, 27872, 21290, 16721, 13458]

In [31]:
cal_large

4788

In [29]:
sum(cal_lst2)/len(owner_small-1)

0.8962148388649801

In [33]:
(cal_large-1)/len(owner_small-1)

0.004684237204213954

We could see that about 89.6% of user only have less than 10 posts. Only about 0.47% users have larger than 100 posts. And the user who has large number, such as larger than 100 posts, company should focus on this user as main target user.

And another question is that where each owner post user come from?

In [34]:
post_tb[['location','id']].groupby(["location"]).count().reset_index().sort_values(by = "id",ascending = False)

location       id
5843                                NAN  2450878
4078                              INDIA   369623
9307                      UNITED STATES   219532
9291                     UNITED KINGDOM   159050
3282                            GERMANY   151113
1705                             CANADA    88059
3097                             FRANCE    84791
9410                                USA    84267
829                           AUSTRALIA    65511
1625                                 CA    61892
5943                        NETHERLANDS    55410
7469                             RUSSIA    45051
9225                                 UK    45002
6953                             POLAND    44033
4307                              ITALY    39085
4278                             ISRAEL    38728
8506                             SWEDEN    34229
8527                        SWITZERLAND    33129
1449                             BRAZIL    30921
8286                              SPAIN    30278
6686                           PAKISTAN    28619
6512                 ON THE SERVER FARM    27030
1972                              CHINA    24297
1142                            BELGIUM    23664
9671                                 WA    23102
9243                            UKRAINE    22681
9163                                 TX    21550
4199                               IRAN    19721
2318                     CZECH REPUBLIC    18816
4235                            IRELAND    18503
844                             AUSTRIA    17877
6366                                 NY    17580
7928                          SINGAPORE    17572
5172                                 MA    17376
6019                        NEW ZEALAND    17241
985                          BANGLADESH    17126
7402                            ROMANIA    17005
2476                            DENMARK    16565
2501                        DEUTSCHLAND    14798
9131                             TURKEY    14325
8152                       SOUTH AFRICA    14314
8315                          SRI LANKA    13666
1670                         CALIFORNIA    13549
6869                        PHILIPPINES    13397
3015                            FINLAND    13279
5035                             LONDON    13162
6267                             NORWAY    13046
3972                                 IL    12878
4369                              JAPAN    12130
1566                           BULGARIA    11864
7024                           PORTUGAL    11551
3024                                 FL    11262
3473                             GREECE    11214
6656                                 PA    10407
2711                              EARTH    10392
962                           BANGALORE    10384
693                           ARGENTINA    10164
3878                            HUNGARY     9101
8863                    THE NETHERLANDS     8899
9585                            VIETNAM     8719
4131                          INDONESIA     8691
6544                                 OR     7847
2838                            ENGLAND     7786
5438                             MEXICO     7674
2799                              EGYPT     7659
6433                                 OH     7494
3193                                 GA     7489
2115                                 CO     7446
5248                           MALAYSIA     7280
2921                             EUROPE     7223
5880                                 NC     6624
5932                              NEPAL     6398
5453                                 MI     6238
1130                            BELARUS     6026
5581                                 MN     5933
...                                 ...      ...
4253                              IRSKO        1
4252                              IRSKA        1
4243                            IRELANS        1
8318                SRI LANKA/SINGAPORE        1
4242                 IRELAND/MANCHESTER        1
4241         

In [35]:
(369623+219532+159050+151113+88059+84267+45002)/5594392

0.1996009575303268

In [36]:
2450878/5594392

0.43809550707208217

NaN is about 43.8% of the users information.

Except for the NaN values, the main user 20% is from India, United States/ USA, United Kingdom/UK, Germany, Canada, France. Compared to the result before, we found that user from Germany is more active than in Canada even though canada registed more user. The other tendency is keep still in this feature.

Here, we come to the tag information of the dataset. We first want to make clear which tag is most popular.

In [39]:
tags_total = []
post_tags = post_tb.tags.tolist()
for each in post_tags:
    tags_total.extend(str(each).split("|"))

In [40]:
from collections import Counter

In [46]:
dic_tags = dict(Counter(tags_total))
sorted(dic_tags.items(), key=lambda item:item[1], reverse=True)

[('nan', 3206012),
 ('javascript', 272130),
 ('java', 229712),
 ('php', 182744),
 ('android', 180511),
 ('c#', 165000),
 ('python', 144287),
 ('jquery', 136671),
 ('html', 125674),
 ('ios', 107391),
 ('css', 88973),
 ('c++', 83529),
 ('mysql', 78722),
 ('angularjs', 73174),
 ('sql', 56465),
 ('swift', 53175),
 ('arrays', 46236),
 ('ruby-on-rails', 42930),
 ('r', 42720),
 ('c', 41255),
 ('json', 40508),
 ('node.js', 40465),
 ('asp.net', 38705),
 ('objective-c', 38306),
 ('sql-server', 34759),
 ('ruby', 29282),
 ('ajax', 28827),
 ('regex', 28793),
 ('excel', 26307),
 ('linux', 26230),
 ('xml', 25360),
 ('asp.net-mvc', 25060),
 ('.net', 24804),
 ('django', 24360),
 ('wordpress', 23084),
 ('spring', 23067),
 ('twitter-bootstrap', 21420),
 ('xcode', 20194),
 ('database', 19702),
 ('string', 19324),
 ('html5', 19269),
 ('vba', 18962),
 ('mongodb', 18217),
 ('wpf', 17967),
 ('windows', 17285),
 ('vb.net', 16734),
 ('eclipse', 16652),
 ('multithreading', 15962),
 ('python-2.7', 15879),
 ('bash

In [47]:
dic_tags = dict(Counter(tags_total))
sorted(dic_tags.items(), key=lambda item:item[1], reverse=True)[1:11]

[('javascript', 272130),
 ('java', 229712),
 ('php', 182744),
 ('android', 180511),
 ('c#', 165000),
 ('python', 144287),
 ('jquery', 136671),
 ('html', 125674),
 ('ios', 107391),
 ('css', 88973)]

We could know that except for the NaN tags, the most popular 10 tags are listed above. javascript takes the first place. Which means the most popular questions. So company could pay more attention to more popular items to get a better business result.

And another thing is that we also want to know for people in different countries, what kind of tag is more popular, i.e. what questions people tend to more posts? this question may help company in improving business model in individual countries.

For example, in India

In [48]:
df_country_tag = post_tb[["location","id","tags"]]
df_country_tag_indian = df_country_tag[df_country_tag["location"]=="INDIA"]
tags_total_india = []
df_country_tag_india_list = df_country_tag_indian.tags.tolist()
for each in df_country_tag_india_list:
    tags_total_india.extend(str(each).split("|"))
dic_tags_india = dict(Counter(tags_total_india))

In [50]:
sorted(dic_tags_india.items(), key=lambda item:item[1], reverse=True)[1:11]

[('android', 14460),
 ('java', 14009),
 ('javascript', 13249),
 ('php', 9595),
 ('jquery', 8256),
 ('c#', 6523),
 ('ios', 6512),
 ('html', 5580),
 ('angularjs', 5256),
 ('python', 4871)]

For example, In United States/USA

In [51]:
df_country_tag_usa = df_country_tag[(df_country_tag["location"]=="UNITED STATES") | (df_country_tag["location"]=="USA")]
tags_total_usa = []
df_country_tag_usa_list = df_country_tag_usa.tags.tolist()
for each in df_country_tag_usa_list:
    tags_total_usa.extend(str(each).split("|"))
dic_tags_usa = dict(Counter(tags_total_usa))

In [52]:
sorted(dic_tags_usa.items(), key=lambda item:item[1], reverse=True)[1:11]

[('javascript', 11319),
 ('java', 7297),
 ('python', 6739),
 ('c#', 5741),
 ('ios', 4986),
 ('android', 4370),
 ('jquery', 4365),
 ('html', 4212),
 ('php', 4160),
 ('angularjs', 3480)]

Except for nan values, the most popular posts in India is android, java, and javascript. In America, is javascript, java, python. In India python is 10th but in America python is 3rd. While in the total result, the first place is taken by javascript,java, php so company should pay more attention to different business model in different countries/regions.

And we want to know the view count information between view count and tags, we have already know the popular tages, take javascript, java

In [72]:
javascript = post_tb.loc[post_tb.tags.str.contains("javascript", na=False)]
java = post_tb.loc[post_tb.tags.str.contains("java", na=False)]

In [73]:
javascript["view_count"].mean()

221.28726464818016

In [74]:
javascript["view_count"].median()

73.0

In [75]:
javascript["view_count"].sum()

60457451.0

In [76]:
javascript["view_count"].std()

900.1724988258874

In [77]:
java["view_count"].mean()

229.8248259837946

In [78]:
java["view_count"].median()

78.0

In [79]:
java["view_count"].sum()

116717687.0

In [80]:
java["view_count"].std()

908.2693175460565

We could see that in the statistical performance different tages performs differently.

And we now have month information, it is still a good method to do research monthly.

We first generate 12 tables and then do some analysis on individual tables.

In [69]:
post_tb['creation_month'] = post_tb['creation_month'].apply(lambda x:int(x))

In [70]:
post_tb[post_tb["creation_month"]==12]

id                                              title  \
0        34016263               Real alternative for Google Feed API   
2        34054617  Magento 2 installing language packs (nl_NL tra...   
7        34498830     Decode X-JWT-Assertion using axiom-api in java   
11       34546446       Flattening TypeScript typings or interfaces?   
68       34074113                                                NaN   
69       34435132                                                NaN   
81       34292059                                                NaN   
88       34041036                                                NaN   
100      34233107                                                NaN   
106      34508989                                                NaN   
130      34409204                                                NaN   
132      34219292                                                NaN   
133      34055321                                                NaN   
136      34071315                                                NaN   
162      34092666                                                NaN   
164      34032750                                                NaN   
165      34298783                                                NaN   
180      34078674                                                NaN   
184      34425921                                                NaN   
188      34193082                                                NaN   
212      34115365                                                NaN   
231      34523759                                                NaN   
232      34337991                                                NaN   
251      34444749                                                NaN   
282      34146201                                                NaN   
299      34228416                                                NaN   
305      34015632                                                NaN   
314      34379876                                                NaN   
319      34432612                                                NaN   
343      34396785                                                NaN   
344      34247246                                                NaN   
346      34153270                                                NaN   
347      34181176                                                NaN   
363      34180600                                                NaN   
370      34282518                                                NaN   
406      34471698                                                NaN   
408      34314583                                                NaN   
421      34084137                                                NaN   
438      34257409                                                NaN   
439      34204048                                                NaN   
451      34458986                                                NaN   
455      34328442                                                NaN   
472      34344515                                                NaN   
476      34062077                                                NaN   
479      34317953                                                NaN   
481      34140245                                                NaN   
483      34379725                                                NaN   
512      34506323                                                NaN   
547      34545388                                                NaN   
558      34515428                                                NaN   
574      34226814                                                NaN   
579      34411988                                                NaN   
596      34243357                                                NaN   
600      34492455                                                NaN   
604      34260992                                      

For example, the information of month 12 is listed above. We could do the similar analysis on it.

We also want to do prediction experiment.

Given that __view_count__ records the number of views the posts have been paid visit to, combining with some other statistics like __answer_count__, __comment_count__ or __favorite_count__, we can obtain normalized indicators to measure the posts with various metrics in different dimensions.
Hereby, we can define the __popularity_index__ as 
$$\textrm{popularity_index} = \frac{ \textrm{favorite_count} } { \textrm{view_count} }$$
where the action of users marking a post as favorite is a good proxy of how useful the user reagrds the post and, thus, how popular the post might be in general. The counts divided by the total number of views make sure the index is normalized within $[0, 1]$, which makes the index comparable for different posts. The __popularity_index__ can be used to rank the post and recommend to general users or categorized by different tags and recommended specifically to users searching for the correponding tags.

Furthermore, we also define the __informativeness_index__ as
$$\textrm{informativeness_index} = \frac{ 0.7 * \textrm{answer_count} + 0.3 * \textrm{comment_count} } {\textrm{view_count}}$$
where the number of answers and comments serves as a good proxy of the completeness of the responses to the post, in other words the richness and quality of the information in the post which can be useful to new users. Here the coefficients are manually assigned to the __answer_count__ and __comment_count__ respectively as 0.7 and 0.3, in order to weigh the answers to the post more important. Similar to the __popularity_index__, the sum of the counts is normalized by the total number of views. This index can be categorized by different tags and ranked in order to make recommendations for users (potentially new and inexperienced) that are searching for answers to certain tags(topics).

Finally, we can also define a composite index __ignorance_index__ as
$$\textrm{ignorance_index} = \frac{\textrm{popularity_index}}{\textrm{informativeness_index}} = \frac{\textrm{favorite_count}}{0.7 * \textrm{answer_count} + 0.3 * \textrm{comment_count}}$$
where we combine the previously defined two indices and show the level of _ignorance_ of the post, which means that the topic/question mentioned in the post is popular and relevant in general, but there lacks informative or complete answers and discussions. Similarly to the __informativeness_index__, this index can be categorized by different tags and ranked in order to make recommendations for users (potentially experienced) that are searching for tags(topics) to share and exchange knowledges.

In [85]:
def cal_ign(x):
    if math.isnan(x['favorite_count']) or math.isnan(x['answer_count']) or math.isnan(x['comment_count']):
        return 0
    else:
        return int(x['favorite_count'])/(0.7*int(x['answer_count'])+0.3*int(x['comment_count']))    

this is the function to do the ignorance calculation